test/integration/integration.test.ts

```diff
@@ -0,0 +1,678 @@
+import { SpaceEventMap } from '../../src/Space.js';
+import { MembersEventMap } from '../../src/Members.js';
+import { CursorsEventMap } from '../../src/Cursors.js';
+import { nanoid } from 'nanoid';
+import { LocksEventMap } from '../../src/Locks.js';
+
+describe.concurrent(
+  'integration tests',
+  () => {
+    it('space members', async (context) => {
+      // Given Spaces clients `performerSpaces` and `observerSpaces`, both configured to use the same API key, and each configured to use a different randomly-generated client ID...
+      const [{ spaces: performerSpaces, clientId: performerClientId }, { spaces: observerSpaces }] =
+        await createClients({ count: 2 });
+
+      // ...and Space instances `performerSpace` and `observerSpace`, fetched from `performerSpaces` and `observerSpaces` respectively using `#get` with a randomly-generated space name, and passing an `offlineTimeout` option of 5 seconds when creating `observerSpace`...
+      //
+      // (Motivation for choosing 5 seconds for `offlineTimeout`: in scenario_1_4 we wait to observe the members `remove` event, which requires us to wait a duration of `offlineTimeout`, and hence means that execution of this test case will take at least `offlineTimeout`. We choose 5 seconds as an arbitrary value which is small enough such that we’re happy for the test to take this long to execute, but not so short as to run the risk of missing this `remove` event whilst waiting for scenario_1_3’s `leave()` to return.)
+      const spaceName = nanoid();
+      const performerSpace = await performerSpaces.get(spaceName);
+      const observerSpace = await observerSpaces.get(spaceName, { offlineTimeout: 5000 });
+
+      // (scenario_1_1: entering a space)
+
+      // (Start of setup for assertions_1_1)
+      const promisesForEventsTriggeredByEntering = {
+        // ...then, when we add a listener to `observerSpace` for the `update` event using `#once`...
+        spaceUpdate: new Promise<SpaceEventMap['update']>((resolve) => {
+          observerSpace.once('update', resolve);
+        }),
+        // ...and add a listener to `observerSpace.members` for the `enter` event using `#once`...
+        //
+        // Note: I think that there are situations in which this listener might not get called, for the same reasons as described in commit 6e77941 — that is, the call to `performerSpace.enter()` might result in `observerSpace` seeing a presence UPDATE instead of an ENTER and hence not emitting a members `enter` event. I haven’t observed this happening in practice yet, and I’d like to be able to test that the SDK emits a members `enter` event, so I’m going to keep the test as it is for now. If we can think of a way for this test to be sure that a members `enter` event will be emitted, then we should implement it.
+        membersEnter: new Promise<MembersEventMap['enter']>((resolve) => {
+          observerSpace.members.once('enter', resolve);
+        }),
```

Comment:
Before I change this, I want to be sure I understand what you're suggesting. Are you saying that you don't want us to use `once` at all in this test file? And, if so, would using `subscribe` and then removing the subscriber when the listener is called — which would emulate the behaviour of `once` and keep the test structure as it is — be OK, or are you wanting us to only call `subscribe` once at the start of the test and then look at which events the listener receives as we perform various actions?

Reply:
(and are you talking only about the members test, or all of them?)

Reply:
no, this was just referring to members, and in particular the comment above and your worry about `present` events. `once` forces you to pass an event name, while you can use `subscribe` to capture everything, including `present` events (which honestly, as I write it, feels like a bug because it's not obvious at all). 

And I proposed `getAll` because it waits for the presence set to sync, I thought would be useful in tests too (as at least in theory, you can use it to expect some state to sync - same as `once`). But I don't mind using `once` for tests - it's still a valid approach.

To be clear, I'm not requesting any changes here, I was responding more to the comment and reference to `enter`/`present` behaviour.